File: Haskell/domaci2.hs
import Data.Char
-- Definisati funkciju koja prima listu brojeva i ukoliko je
-- ona parne duzine, svaki element kvadrira, inace svaki element
-- mnozi sa 10. Koristiti funkciju map i lambda funkcije.
funkcija1 [] = []
funkcija1 l
    | mod (length l) 2 == 0 = map (\x -> x*x) l
    | otherwise = map (\x -> x*10) l

-- Napisati funkciju koja prima niz karaktera. Prvo je potrebno
-- funkcijom filter izbaciti sva velika slova, a zatim pomocu funkcije
-- map pretvoriti sva preostala slova u velika.
-- Napomena: Potrebno je u .hs fajl importovati funkciju toUpper
-- iz modula Data.Char, na sledeci nacin:
-- import Data.Char -- Dodati na pocetak fajla
funkcija2 :: [Char] -> [Char]
funkcija2 l = map (toUpper) (filter (\x -> (x >= 'a' && x <= 'z')) l)

-- Napisati funkciju "primeni" koja prima 3 parametra:
--	1) f1 :: [[Int]] -> [[Int]] - funkciju koja prihvata [[Int]] i vraca [[Int]]
--	2) f2 :: [[Int]] -> [Int] - funkciju koja prihvata [[Int]] i vraca [Int]
--	3) listu ciji su elementi liste brojeva, tj. [[Int]]
--	Funkcija treba da primeni funkcije f1 i f2 nad prosledjenom listom
--	i vrati rezultat (listu tipa [[Int]] pretvara u [Int], pomocu 
--	funkcija f1 i f2).
primeni f1 f2 l = f2 $ f1 l

-- Definisati funkciju "izbaciParnePaSumiraj" preko funkcije "primeni"
--   i funkcija iz zadataka 2 i 3 (sa Vezbi 2), koristeci parcijalnu 
-- primenu funkcija.

-- Definisati funkciju prosecnaDuzina koja prima niz stringova
-- i racuna njihovu prosecnu duzinu pomocu funkcije fold i map.
-- Dovoljno je vratiti celobrojnu vrednost (npr. 5 umesto 5.3)
prosecnaDuzina l = div suma (length l)
    where duzine = map length l
          suma = foldr (+) 0 duzine

File: Scala/Domaci2/Domaci2.scala
/*
1. Definisati funkciju koja prima listu brojeva i ukoliko je
   ona parne duzine, svaki element kvadrira, inace svaki element
   mnozi sa 10. Koristiti funkciju map.

2. Napisati funkciju koja prima niz karaktera. Prvo je potrebno
   funkcijom filter izbaciti sva velika slova, a zatim pomocu funkcije
   map pretvoriti sva preostala slova u velika.

3. Definisati funkciju prosecnaDuzina koja prima niz stringova
   i racuna njihovu prosecnu duzinu pomocu funkcije reduce i map.

4. Definisati funkciju koja prima niz brojeva. Zadrzi one brojeve
   kojima je prva i poslednja cifra ista, te svaki broj "okrene"
   po principu: "12341" -> "14321". Koristiti map i filter.

5. Definisati funkciju koja prima niz stringova. Treba da izbaciti
   sve one koji imaju vise od 4 velika slova. Nakon toga potrebno
   je na svaki od njih nalepiti sebe u "ogledalu" ("abc" -> 
   "abccba") i na kraju ih sve spojiti u jedan veliki string. 
   Koristiti map, filter, reduce.
*/


object Domaci2 {
    def main(args: Array[String]) = {
        // var l = prvi(List(1,2,3,4,5,6))
        // for(i <- l) println(i.toString + " ")

        // var l = drugi(Array('a','B','c'))
        // for(i <- l) println(i.toString + " ")

        //println(prosecnaDuzina(Array("aaaa", "ababababb", "")))

        //print(cetvrti(Array(123456,1234233331,525235,231435252)).toList)

        print(peti(Array("Marko", "DAniLOS", "stevan", "Miodrage")))
    }

    def peti(niz : Array[String]) = {
        def velika4(s : String) : Boolean = {
            var i = 0
            for(c <- s.toCharArray)
                if(c.isUpper)
                    i += 1
            
            if(i > 4)
                false
            else 
                true
        }
        niz.filter(velika4).map(x => x + x.reverse).reduce(_ + _)
    }

    def cetvrti(niz : Array[Int]) : Array[Int] = {
        def ist(x : Int) : Boolean = {
            if(x < 10)
                true
            else {
                var pos = x % 10
                var n = x
                while(n > 9){
                    n /= 10
                }
                n == pos
            }
        }

        def okreni(x : Int) : Int = {
            var ok = 0
            var n = x
            while(n > 0){
                ok = ok * 10 + n % 10
                n /= 10
            }
            ok
        }

        niz.filter(ist).map(okreni)
    }



    def prosecnaDuzina(niz : Array[String]) : Double = {
        var n = niz.length
        var s = niz.map(x => x.length).reduce(_ + _)
        s / (1.0 * n)
    }

    def drugi(nc : Array[Char]) : Array[Char] = {
        nc.filter(x => x.isLower).map(x => x.toUpper)
    }

    def prvi(l : List[Int]) : List[Int] = {
        if(l.length % 2 == 0){
            return l.map(x => x * x)
        } else {
            return l.map(x => x * 10)
        }
    }
}